frame/scheduler/src/lib.rs

```diff
@@ -346,50 +445,64 @@ impl<T: Trait> Module<T> {
 		Ok(address)
 	}
 
-	fn do_cancel_named(id: Vec<u8>) -> Result<(), DispatchError> {
-		if let Some((when, index)) = Lookup::<T>::take(id) {
-			let i = index as usize;
-			Agenda::<T>::mutate(when, |agenda| if let Some(s) = agenda.get_mut(i) { *s = None });
-			Self::deposit_event(RawEvent::Canceled(when, index));
-			Ok(())
-		} else {
-			Err(Error::<T>::FailedToCancel)?
-		}
+	fn do_cancel_named(origin: Option<T::PalletsOrigin>, id: Vec<u8>) -> DispatchResult {
+		Lookup::<T>::try_mutate_exists(id, |lookup| -> DispatchResult {
+			if let Some((when, index)) = lookup.take() {
+				let i = index as usize;
+				Agenda::<T>::try_mutate(when, |agenda| -> DispatchResult {
+					if let Some(s) = agenda.get_mut(i) {
+						if let (Some(ref o), Some(ref s)) = (origin, s.borrow()) {
+							if *o != s.origin {
+								return Err(BadOrigin.into());
+							}
+						}
+						*s = None;
+					}
+					Ok(())
+				})?;
+				Self::deposit_event(RawEvent::Canceled(when, index));
+				Ok(())
+			} else {
+				Err(Error::<T>::FailedToCancel)?
+			}
+		})
 	}
 }
 
-impl<T: Trait> schedule::Anon<T::BlockNumber, <T as Trait>::Call> for Module<T> {
+impl<T: Trait> schedule::Anon<T::BlockNumber, <T as Trait>::Call, T::PalletsOrigin> for Module<T> {
 	type Address = TaskAddress<T::BlockNumber>;
 
 	fn schedule(
 		when: T::BlockNumber,
 		maybe_periodic: Option<schedule::Period<T::BlockNumber>>,
 		priority: schedule::Priority,
+		origin: T::PalletsOrigin,
 		call: <T as Trait>::Call
 	) -> Self::Address {
-		Self::do_schedule(when, maybe_periodic, priority, call)
+		Self::do_schedule(when, maybe_periodic, priority, origin, call)
 	}
 
 	fn cancel((when, index): Self::Address) -> Result<(), ()> {
-		Self::do_cancel((when, index)).map_err(|_| ())
+		Self::do_cancel(None, (when, index)).map_err(|_| ())
```

Comment:
I'm not sure about this:

cancelling a scheduled requires giving the origin if pallet-scheduler is used by its module directly, but if pallet-scheduler is used throught Anon trait then it doesn't require the origin to be given anymore, instead it just bypass this check.

This means if a pallet use Anon trait every origin accepting to cancel a scheduled can cancel every possible schedule even those scheduled through other origin. It seems not correct and that instead origin must be check.
(or otherwise we shouldn't check the origin in both case at least)

Reply:
The calls can be dispatched from extrinsic, and in my use case, we have multiple councils and each of them have different origin, and they shouldn't be able to cancel other's scheduled call. So we have to check origin in cancel.

A method from trait is implicitly trusted as it can only be called from runtime itself. e.g. `trait Currency` is able to mint without restrictions. So I don't see this is incorrect per se. It is up to the caller to check the origin does indeed have the permission to cancel the call.

But I can see your point, lack or origin argument could be an issue for some use cases. e.g. replicate the cancel dispatchable call. At the same time I am hesitated to add `origin: T::PalletsOrigin` as it will also add usability issue to cancel task (will need to access the correct `origin`, but we do not expose a getter to access it. Maybe add `origin: Option<T::PalletsOrigin>` or have two distinct cancel methods?

Reply:
Hmm I see

> At the same time I am hesitated to add origin: T::PalletsOrigin as it will also add usability issue to cancel task (will need to access the correct origin, but we do not expose a getter to access it.

At some point currently this is also currently not an issue: democracy will always use Root to cancel as well.

But I agree, with all statements, I'm ok with what this PR has currently then.